### ISSUE TITLE
Pretrained weights

### ISSUE THREAD

**Issue by saurabbhsp** [jackd/template_ffd#15]
Hi,
I am looking for pretrained weights for firearm category for comparison purpose. Is it possible  to get access to pretrained weights.
I tried to re implement the paper and train the pistol/firearm category and the final loss was at 15 with batch size of 32. However the test deformations were not satisfactory.

**Comment by jackd** [comment]
I'm run off my feet for the next few days but I'll get back to this next week hopefully. One thing I've realized since writing this paper is I mis-interpreted the category name and used pistols instead of rifles, which I compared against... I think. It's been a while, memory is a bit hazy. If you're seeing highly unstaisfactory results and it's urgent I'd suggest you double check the templates for the categories are close to the examples in the category - i.e. if you're trying to deform pistols into rifles things may not go so well...

**Comment by saurabbhsp** [comment]
Sure, I am working with pistol category for training and testing. This is not that urgent, get back to me when you are free.


Thanks for all the help

**Comment by jackd** [comment]
Hi Saurabh,
Can you try visualizing your results with [this line](https://github.com/jackd/template_ffd/blob/master/model/template_ffd_builder.py#L76) hard-coded to `training = True` ? keras introduced some braking changes in a subsequent release that meant model update ops no longer got added to the global collection, so it's possible your network training didn't update batch norm statistics appropriately.

I want to do a major refactoring of this code-base shortly which will remove this issue, but if there's something more sinister at foot I'd rather know about it beforehand :)

**Comment by saurabbhsp** [comment]
Thanks for your help.

Yes, I tried to train with this. I also calculated the chamfer report(Normalized) for pistol category and the loss was around 72. (1000 * chamfer). I am not sure if this is in expected range of error as the paper has reported the error for pistol category as 39.

**Comment by jackd** [comment]
Interesting. I can't recall doing anything particularly fancy with pistols. Just cleaning up a few life things since deadline has passed before I can get to this, but in the mean time: can you view the loss in tensorboard? Is the evaluation loss significantly different to the training loss?

**Comment by saurabbhsp** [comment]
For training the loss was flatlined to values between 14-17 and test loss was at 28. I will also try another category. I was thinking this may be due to random train and test datasets. I should perform multiple runs of the code.
Also the numbers reported in paper, are the using bugged version or non bugged version.

Another observation is that the point clouds generated using 1024 points is much better representation than the mesh generated using default points.

**Comment by saurabbhsp** [comment]
I think I got my mistake with reconstruction for meshes. As I mentioned before
1) I have very good quality point clouds with 1024 vertices (Inferred for templates)
2) Poor mesh reconstruction. This was not done with rich vertices and faces but with the default template vertices.

Could it be that this is root cause for the poor reconstructions. If so how to generate faces in denser point cloud representations with 1024 points.


Edit:- I forgot the edge length threshold discussed in the paper ie 0.02. I think this will solve the issue.

Thanks for all the help !!

**Comment by jackd** [comment]
Yeah, that will give you issues :). It had me stumped for a while when I was visualizing point clouds and looking at IoU metrics calculated from the mesh. Feel free to re-open if this doesn't resolve it.
